pkg/merkle/proof.go

```diff
@@ -165,173 +127,140 @@ type nodeQueue struct {
-			lowerBound >>= 1
+		// If we reach the sub-root (i.e. `index == 1`), we can return the root (i.e. `index == 0`) by
+		// hashing the tree's leaf count with the last computed hash.
+		if nodeIdx == 1 {
+			return HashRoot(leafCount, queue[(writeIdx-1)%queueLen].hash), nil
 		}
 
-		if isLeftChild(current.index) {
-			// Handle left-child branch.
-			left = current.hash
-			if head < len(queue) && queue[head].index == current.index+1 {
-				right = queue[head].hash
-				head++
-			} else {
-				right, err = p.getNextProof(&proofIdx)
-				if err != nil {
-					return nil, err
-				}
-			}
-		} else {
-			// Handle right-child branch.
-			right = current.hash
-			if head < len(queue) && queue[head].index == current.index-1 {
-				left = queue[head].hash
-				head++
-			} else {
-				left, err = p.getNextProof(&proofIdx)
-				if err != nil {
-					return nil, err
-				}
+		// If node index we are handling is the upper bound and is even, then it's sibling to the right does not
+		// exist (since this is an unbalanced tree), so we can just copy the hash up one level.
+		if nodeIdx == upperBound && isEven(nodeIdx) {
+			queue[writeIdx%queueLen] = nodeQueue{
+				index: nodeIdx >> 1,
+				hash:  HashPairlessNode(queue[readIdx%queueLen].hash),
 			}
-		}
+			writeIdx++
+			readIdx++
 
-		queue = append(queue, nodeQueue{
-			index: current.index >> 1,
-			hash:  HashNode(left, right),
-		})
-	}
-
-	return nil, ErrNilRoot
-}
+			// If we are not at the lowest tree index (i.e. there are nodes to the left that we have yet to process at
+			// this level), then continue.
+			if nodeIdx != lowestTreeIndex {
+				continue
+			}
 
-// makeIndexedValues creates indexed values from elements and their indices.
-func makeIndexedValues(leaves []Leaf, indices []int) (IndexedValues, error) {
-	result := make(IndexedValues, len(indices))
+			// If we are at the lowest tree index (i.e. there are no nodes to the left that we have yet to process at
+			// this level), then we can update the lower bound and upper bound for the next level up.
+			lowestTreeIndex >>= 1
+			upperBound >>= 1
 
-	for i, idx := range indices {
-		result[i] = IndexedValue{
-			value: leaves[idx],
-			index: idx,
+			continue
 		}
-	}
-
-	return result, nil
-}
 
-// makeIndices returns a slice of ascending ordered indices for the given starting index and count.
-func makeIndices(startingIndex, count int) ([]int, error) {
-	if startingIndex < 0 || count <= 0 {
-		return nil, ErrInvalidRange
-	}
-
-	indices := make([]int, count)
-	for i := 0; i < count; i++ {
-		indices[i] = startingIndex + i
-	}
-
-	return indices, nil
-}
+		nextNodeIdx := queue[(readIdx+1)%queueLen].index
 
-// validateIndices validates the indices slice of a proof.
-func validateIndices(indices []int, leafCount int) error {
-	balancedLeafCount, err := CalculateBalancedNodesCount(leafCount)
-	if err != nil {
-		return err
-	}
+		var (
+			right []byte
+			left  []byte
+		)
```

Comment:
Avoid recreating right and left every iteration by moving them out of the for loop.